### PR TITLE
Add scripts to easily run a-c and fenix smoketests.

### PR DESCRIPTION
**Pull request by rfk**
Like the scripts that @eoger added recently for iOS, but for android.

The fenix script depends on https://github.com/mozilla-mobile/fenix/pull/8681 which hasn't merged to master yet.

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - `automation/all_tests.sh` runs to completion and produces no failures
  - Note: For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/master/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: automation/shared.py [hunk 1]
```diff
@@ -1,29 +1,64 @@
...
     print(f"> \033[34m{msg}\033[0m")
 
 def fatal_err(msg):
     print(f"\033[31mError: {msg}\033[0m")
     exit(1)
 
-# Run command and let subprocess throw an exception when
-# a command returns a non-zero status.
 def run_cmd_checked(*args, **kwargs):
+    """Run a command, throwing exception if it exits with non-zero status."""
```
Comment (eoger) [inline automation/shared.py]: Nice 👍

File: automation/shared.py [hunk 2]
```diff
@@ -1,29 +1,64 @@
...
 
-# Ensure there is no un-commited or staged files in the working tree.
 def ensure_working_tree_clean():
+    """Error out if there are un-committed or staged files in the working tree."""
     if run_cmd_checked(["git", "status", "--porcelain"], capture_output=True).stdout:
         fatal_err("The working tree has un-commited or staged files.")
 
-# Find the absolute path to the Application Services repository root.
 def find_app_services_root():
+    """Find the absolute path o the Application services repository root."""
```
Comment (eoger) [inline automation/shared.py]:
```suggestion
    """Find the absolute path of the Application services repository root."""
```

File: automation/shared.py [hunk 3]
```diff
@@ -1,29 +1,64 @@
...
     cur_dir = Path(__file__).parent
     while not Path(cur_dir, "LICENSE").exists():
         cur_dir = cur_dir.parent
     return cur_dir.absolute()
+
+def set_gradle_substitution_path(project_dir, name, value):
+    """"Set a substitution path property in a gradle `local.properties` file.
```
Comment (eoger) [inline automation/shared.py]:
```suggestion
    """Set a substitution path property in a gradle `local.properties` file.
```

File: automation/smoke-test-android-components.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+if repo_path is None:
+    repo_path = tempfile.mkdtemp(suffix="-a-c")
+    if remote_repo_url is None:
+        remote_repo_url = DEFAULT_REMOTE_REPO_URL
+    step_msg(f"Cloning {remote_repo_url}")
+    run_cmd_checked(["git", "clone", remote_repo_url, repo_path])
+    if branch is not None:
+        run_cmd_checked(["git", "checkout", branch], cwd=repo_path)
+elif branch is not None:
+    fatal_err("Cannot specify fenix branch when using a local repo; check it out locally and try again.")
```
Comment (eoger) [inline automation/smoke-test-android-components.py]: Would you mind replicating this behavior (no default value for `branch`) in the iOS script for consistency?

**Comment by rfk** [comment]
> Do we want to add a line or two in https://github.com/mozilla/application-services/blob/master/docs/howtos/smoke-testing-app-services.md to document these scripts?

Yes, good catch, thanks.

File: docs/howtos/smoke-testing-app-services.md [hunk 5]
```diff
@@ -7,4 +7,15 @@ It can be done manually using substitution scripts, but we also have scripts tha
...
 The `automation/smoke-test-firefox-ios.py` script will clone (or use a local version) of Firefox iOS and
 run tests against the current application-services worktree.  
-Add the `-h` argument to discover all the script exciting options!
+Add the `-h` argument to discover all of the script's exciting options!
+
+## Android Components
+
+The `automation/smoke-test-android-components.py` script will clone (or use a local version) of
+android-components and run a subset of its tests against the current application-services worktree.
+It tries to only run tests that might be relecant to application-services functionality.
```
Comment (eoger) [inline docs/howtos/smoke-testing-app-services.md]: Nit
```suggestion
It tries to only run tests that might be relevant to application-services functionality.
```

Comment (rfk) [reply docs/howtos/smoke-testing-app-services.md]: I think my typos are getting worse the older I get...

**Pull request merged**